ARSN-401: Add PRIMARY communication in cluster RPC

BourgoisMickael opened the pull request:
Adds:

- Communication with `PRIMARY` only
- Propagate werelogs config (logLevel) from dependant project
- `errorCode` in error communication (for HTTP)

This is used by scuba https://github.com/scality/scuba/pull/79

BourgoisMickael committed:
ARSN-401: Add PRIMARY communication in cluster RPC

bert-e commented:
# Hello bourgoismickael,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).





*Status report is not available.*

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

BourgoisMickael commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/ARSN-401-cluster-rpc-primary` into
`development/7.70`
* [w/8.1/improvement/ARSN-401-cluster-rpc-primary](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/improvement/ARSN-401-cluster-rpc-primary) will be merged into `development/8.1` 



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

BourgoisMickael committed:
ARSN-401: Fix werelogs config in cluster RPC

Also note that there are some arsenal modules that
have some side effect by being imported as they reconfigure
the werelogs logLevel.
Like: lib/storage/data/external/GCP/GcpUtils.js

BourgoisMickael commented:
reset

BourgoisMickael committed:
ARSN-401: Bump werelogs for types

BourgoisMickael commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/improvement/ARSN-401-cluster-rpc-primary` with contents from `improvement/ARSN-401-cluster-rpc-primary`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/improvement/ARSN-401-cluster-rpc-primary origin/development/8.1
 $ git merge origin/improvement/ARSN-401-cluster-rpc-primary
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/improvement/ARSN-401-cluster-rpc-primary
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

williamlardier reviewed (commented):
LGTM but I was expecting some unit tests, can we add them?

jonathan-gramain commented on lib/clustering/ClusterRPC.ts:
```diff
@@ -74,6 +74,7 @@ const rpcLogger = new werelogs.Logger('ClusterRPC');
 
 export type ResultObject = {
     error: Error | null;
+    errorCode?: number;
```
I'd prefer having the error code part of the `error` object itself to keep it tidy, where the `Error` can be reconstructed from something serialized. I'm pretty sure the `Error` type accepts a `code` attribute. The current serialization seems to be a simple string now but it could be instead an object with two fields for example (say `code` and `message`) if the goal is to keep the description of the error along with the code.

The serialization/deserialization should stay internal to ClusterRPC and not exposed to clients so it should be fine to change how it's made.

BourgoisMickael replied:
I moved the code to the `error`

jonathan-gramain reviewed (commented):
LGTM with one comment about the error extension.

I also agree with William that there should be some tests, you can have a look at the test server that is spawned as part of the functional tests and add one or two test cases there. During development you can just start the server locally with a custom port and use `curl` to trigger the tests (each test is declared in the server https://github.com/scality/Arsenal/blob/development/7.70/tests/functional/clustering/ClusterRPC-test-server.js with a name and is invoked from the test code via an HTTP request URL containing the test name). Once it works you can integrate the test case to the CI in https://github.com/scality/Arsenal/blob/development/7.70/tests/functional/clustering/ClusterRPC.spec.js#L64

BourgoisMickael committed:
ARSN-401: add errorCode in cluster RPC for scuba

BourgoisMickael committed:
ARSN-401: Test new RPC communication

bert-e commented:
# History mismatch



Merge commit #d9f60deb1a92aa8c05a1debf1f13b609dd5021e5 on the integration branch
`w/8.1/improvement/ARSN-401-cluster-rpc-primary` is merging a branch which is neither the current
branch `improvement/ARSN-401-cluster-rpc-primary` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `improvement/ARSN-401-cluster-rpc-primary` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

BourgoisMickael commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/improvement/ARSN-401-cluster-rpc-primary` with contents from `improvement/ARSN-401-cluster-rpc-primary`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/improvement/ARSN-401-cluster-rpc-primary origin/development/8.1
 $ git merge origin/improvement/ARSN-401-cluster-rpc-primary
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/improvement/ARSN-401-cluster-rpc-primary
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

BourgoisMickael commented on tests/functional/clustering/ClusterRPC.spec.js:
```diff
@@ -46,13 +46,45 @@ function stopTestServer(done) {
+                responseBody += chunk;
+            })
             .on('end', () => {
-                expect(res.statusCode).toEqual(200);
-                cb();
```
This would cause an exception to be thrown in a callback, that makes an `uncaughtException`, the `cb` is never called and jest hangs a long time until timeout.

BourgoisMickael commented on tests/functional/clustering/ClusterRPC.spec.js:
```diff
@@ -46,13 +46,45 @@ function stopTestServer(done) {
+ * @throws {assert.AssertionError}
+ * @returns {undefined}
+ */
+function handleAssertionError(responseBody) {
```
It produces this output, pointing at the actual assertion from the other file:

```js
  ● ClusterRPC › worker to primary › should succeed and return a result

    assert.strictEqual(received, expected)

    Expected value to strictly be equal to:
      2
    Received:
      1
    
    Message:
      I want to fail

      231 |
      232 |     const results = await sendWorkerCommand('PRIMARY', 'echoHandler', uids, payload);
    > 233 |     assert.strictEqual(1, 2, 'I want to fail');
          |            ^
      234 |     assert.strictEqual(results.length, 1, 'There is 1 and only 1 primary');
      235 |     assert.ifError(results[0].error);
      236 |     assert.deepStrictEqual(results[0].result, expected);

      at Object.workerToPrimaryEcho [as /worker-to-primary/echo] (tests/functional/clustering/ClusterRPC-test-server.js:233:12)

```

BourgoisMickael committed:
ARSN-401: Test clusterRPC fix error response code

In case a regular error without code is thrown

BourgoisMickael commented on tests/functional/clustering/ClusterRPC.spec.js:
```diff
@@ -46,13 +46,45 @@ function stopTestServer(done) {
+                    expect(res.statusCode).toEqual(200);
+                } catch (err) {
+                    if (!(err instanceof assert.AssertionError)) {
+                        err.message += `\n\nBody:\n${responseBody}`;
```
Produces:
```js
Expected: 200
Received: 400

Body:
MissingParameter
```

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-401 contains:


* `7.70.27`

* `8.1.125`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.28`

* `8.1.126`


Please check the `Fix Version/s` of ARSN-401, or the target
branch of this pull request.




*The following options are set:* **create_integration_branches**

anurag4DSB reviewed (approved):
Scope of review: Reviewed from code point of view, still understanding MD
LGTM.
This technique is very interesting. Correct me if I understood wrong. This is useful in scenarios where a library or module (arsenal in this case) needs to accept a configuration object that is meant to be passed to another library or module (werelogs). By directly linking the type of the configuration object to the expected parameter type of the werelogs.configure method, it helps to ensure that the arsenal library will correctly propagate the configuration object to werelogs, even if future versions of werelogs change the expected configuration object structure.

BourgoisMickael committed:
ARSN-401: Bump version

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARSN-401-cluster-rpc-primary` and the
destination branch `development/7.70`.

Please resolve the conflict on **the feature branch** (`improvement/ARSN-401-cluster-rpc-primary`).

```sh
git fetch && \
git checkout origin/improvement/ARSN-401-cluster-rpc-primary && \
git merge origin/development/7.70
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARSN-401-cluster-rpc-primary
```







*The following options are set:* **create_integration_branches**

BourgoisMickael committed:
Merge branch 'development/7.70' into improvement/ARSN-401-cluster-rpc-primary

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.1/improvement/ARSN-401-cluster-rpc-primary` with contents from `improvement/ARSN-401-cluster-rpc-primary`
and `development/8.1`.

Please resolve the conflict on **the integration branch** (`w/8.1/improvement/ARSN-401-cluster-rpc-primary`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.1/improvement/ARSN-401-cluster-rpc-primary
 $ git pull  # or "git reset --hard origin/w/8.1/improvement/ARSN-401-cluster-rpc-primary"
 $ git merge origin/development/8.1
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/ARSN-401-cluster-rpc-primary
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/improvement/ARSN-401-cluster-rpc-primary
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

BourgoisMickael commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`



Please check the status of the associated issue ARSN-401.

Goodbye bourgoismickael.

jonathan-gramain commented:
> This is useful in scenarios where a library or module (arsenal in this case) needs to accept a configuration object that is meant to be passed to another library or module (werelogs). By directly linking the type of the configuration object to the expected parameter type of the werelogs.configure method, it helps to ensure that the arsenal library will correctly propagate the configuration object to werelogs, even if future versions of werelogs change the expected configuration object structure.

My understanding is that, but I may not have the full picture: if we fork a worker before we configure werelogs, we will need to apply the werelogs configuration to both the primary and the worker(s), because they're now different processes. But if we apply the config prior to the fork, then the worker will be already configured since the worker process memory maps the primary's memory (and copy-on-write when written to).
